[RFC] Handle *some* out-of-spec behaviour explicitly

Sopel97 opened the pull request:
This is a preliminary set of changes to explicitly handle some cases that resulted in undefined behaviour previously:

1. UCI is given a fen that has more than 32 pieces
2. UCI is given a fen that has invalid number of kings
3. UCI is given a fen that has a piece configuration definitely unreachable piece counts (prevents positions with >256 moves)
4. UCI is given a fen that would write outside of a board (due to implementation details)
5. UCI is given a fen that otherwise fails `pos_is_ok` check
6. UCI is given a fen with rule50 counter >100

Additional validation changes:

1. During FEN parsing, error on invalid castling rights (previously ignored)
2. During FEN parsing, error on invalid ep-square (previously ignored)
3. During FEN parsing, error on invalid side to move (previously assumed white)
4. During FEN + moves parsing, error on illegal move (previously was being ignored)

TODO (at least what I have on mind right now, there's definitely more):

1. When parsing FEN, ensure that the file cannot change by means other than '/'

Any above errors are handled by calling `std::terminate`. The motivation for such behaviour is lack of specification from the UCI protocol. UCI allows ignoring errors, but ignoring errors never a good idea as it leads to differences between expected and actual state. Crashing is the only responsible thing we can do in this case.

The test under https://tests.stockfishchess.org/tests/view/645b9b0a175801c38d5c4e85 was mostly a sanity check (with a slightly older version), I don't expect any performance degradation. The current state of this PR is that the changes are slapped on top of whatever was there. It may be better to rewrite some parts completely, but first I'd like to know if this is desired, and if I'm not missing something important.

noobpwnftw commented:
> but first I'd like to know if this is desired, and if I'm not missing something important.

No, for two reasons, 1) to the professional intensive care unit people, your `abort()` is equivalent to a segfault, so none of this solves their problem(until they write to agree that we can perform a controlled crash with invalid inputs now), 2) before, people can yolo through some of the less harmful invalids without any problem, now they can't and they will anyways never figure out how to write a proper sanitizer, so in a sense this just made it more punishing to the noobs.

You did not miss anything "important", as it is obvious that it is nothing complicated to filter out positions that will crash SF, however it is complicated to define what is allowed afterwards, do you allow for anything that passes your filter, or is it still "legal positions reachable from startpos"? If the former, are the conditions subject to change or permanent?

peregrineshahin commented on src/position.cpp:
```diff
@@ -214,13 +216,41 @@ Position& Position::set(const string& fenStr, bool isChess960, StateInfo* si, Th
+        UCI::critical_error(std::string("Invalid FEN. Invalid piece: ") + std::string(1, token));
   }
 
+  int pawns_w = pos.count<PAWN>(WHITE);
```
'pos' seems undeclared.

peregrineshahin commented on src/position.cpp:
```diff
@@ -276,6 +317,9 @@ Position& Position::set(const string& fenStr, bool isChess960, StateInfo* si, Th
   // 5-6. Halfmove clock and fullmove number
   ss >> std::skipws >> st->rule50 >> gamePly;
 
+  if (st->rule50 > 100)
```
I think this is hardly an issue since a user can work around this and submit a fen of counter 99, 2 moves earlier and then get his way through to the same fen we rejected.
On a different note If this is really relevant I think what should reject is negative values for integers instead.

vdbergh commented:
> 
> Any above errors are handled by calling `std::terminate`. The motivation for such behaviour is lack of specification from the UCI protocol. UCI allows ignoring errors, but ignoring errors never a good idea as it leads to differences between expected and actual state. Crashing is the only responsible thing we can do in this case.
>

You are not ignoring the error if you send an `info string` message to the user.

The fact that GUIs can currently not parse such a message is a valid concern. But I am sure that if SF commits to a particular error message, GUIs using SF will start to parse it. Then others engines will follow the lead.

The UCI protocol is not totally immutable and can evolve. For example UCI_Variant has been added in this way.

Sopel97 commented:
> before, people can yolo through some of the less harmful invalids without any problem, now they can't

Why do we suddenly care about usage that violates preconditions?

> You did not miss anything "important", as it is obvious that it is nothing complicated to filter out positions that will crash SF, however it is complicated to define what is allowed afterwards, do you allow for anything that passes your filter, or is it still "legal positions reachable from startpos"? If the former, are the conditions subject to change or permanent?

Crashing aside, this PR's purpose is to ensure there is no differences in observed and assumed state of the engine by the user in the 3rd case - where sometimes input is ignored and sometimes everything can happen.

Sopel97 commented:
> The fact that GUIs can currently not parse such a message is a valid concern. But I am sure that if SF commits to a particular error message format, GUIs using SF will start to parse it. Then others engines will follow the lead.

Are you willing to specify the behaviour and state guarantees for every such case, for the other engines and GUIs to implement?

noobpwnftw commented:
> Why do we suddenly care about usage that violates preconditions?

Which precondition? I'm kinda lost in the chain of cause and consequences.

Sopel97 commented:
> > Why do we suddenly care about usage that violates preconditions?
> 
> Which precondition? I'm kinda lost in the chain of cause and consequences.

The precondition that positions passed to the engine must be reachable from startpos

noobpwnftw commented:
> The precondition that positions passed to the engine must be reachable from startpos

And why are we starting to care about usage that violates it?

Sopel97 commented:
> > The precondition that positions passed to the engine must be reachable from startpos
> 
> And why are we starting to care about usage that violates it?

you suggested such a use case as a reason to dismiss this PR

noobpwnftw commented:
> you suggested such a use case as a reason to dismiss this PR

If you argue that the outcome of such violations of precondition is not of concern, then I think we also won't need this PR to deal with the outcome of such violations of precondition. Whether it resulted in "happens to work" or "crashing" is irrelevant to the context.

vdbergh commented:
> > The fact that GUIs can currently not parse such a message is a valid concern. But I am sure that if SF commits to a particular error message format, GUIs using SF will start to parse it. Then others engines will follow the lead.
> 
> Are you willing to specify the behaviour and state guarantees for every such case, for the other engines and GUIs to implement?

I am not getting your point. SF could emit a message like
```
info string Invalid Position. <further expanations>
```
The GUI could recognize the "Invalid Position" part and inform the user of the error (possibly with the further explanations). In that way SF and a compliant GUI would not get out of sync.

This is BTW the way the xboard protocol handles it.

Sopel97 commented:
> > > The fact that GUIs can currently not parse such a message is a valid concern. But I am sure that if SF commits to a particular error message format, GUIs using SF will start to parse it. Then others engines will follow the lead.
> > 
> > 
> > Are you willing to specify the behaviour and state guarantees for every such case, for the other engines and GUIs to implement?
> 
> I am not getting your point. SF could emit a message like
> 
> ```
> info string Invalid Position. <further expanations>
> ```
> 
> The GUI could recognize the "Invalid Position" part and inform the user of the error (possibly with the further explanations). In that way SF and a compliant GUI would not get out of sync.
> 
> This is BTW the way the xboard protocol handles it.

And what happens? What's the state after the error?

XInTheDark commented:
> And what happens? What's the state after the error?

I would presume that the program gracefully exits, or at least this is the most accepted opinion in #4558. To the GUI and user itself it won't make much of a difference though.

ddobbelaere commented:
The best way to change the UCI protocol is to set the tone and return `info error INVALID_FEN`. *Maybe* everyone else follows :)

vdbergh commented:
> The best way to change the UCI protocol is to set the tone and return `info error INVALID_FEN`. _Maybe_ everyone else follows :)

I think it is not a good idea to make an incompatible change. The GUIs will ignore the `info error` message (as it is not valid UCI) and so no one will see it.

vondele commented:
As I noted elsewhere, we already have similar actions relating to the net:
https://github.com/official-stockfish/Stockfish/blob/65e2150501b87e6ce00fae4e3f056444f39462fd/src/evaluate.cpp#L140

TheBlackPlague commented:
> Crashing aside, this PR's purpose is to ensure there is no differences in observed and assumed state of the engine by the user - where sometimes input is ignored and sometimes everything can happen.

To be very honest, I agree that removing undefined behavior is a good idea. I'm for this proposed change.

However, might I suggest this be released as part of a major release, being a backwards compatibility breaking change (it can't be considered as a bug fix)? I'm not sure if Stockfish even cares about semantic versioning — but it's a nice thing for the end users.

Sopel97 commented:
Stockfish doesn't use semantic versioning so there's no way to indicate a breaking change like that.

Disservin commented:
why not throw a runtime error?

TheBlackPlague commented:
> why not throw a runtime error?

You mean without crashing?

Disservin commented:
> > why not throw a runtime error?
> 
> You mean without crashing?

`throw std::runtime_error("Unexpected fen, failed to parse: ....");`

kills the program unless a try and catch block surrends it.

Disservin commented:
> > why not throw a runtime error?
> 
> stockfish doesn't use exceptions, and without exceptions you get std::terminate

if that is how it is then im fine with it. Though perhaps throwing exceptions might give the user more detail about the error that occured.  It might be that the user is not able to see the uci info because the gui hide's it but exception's might be shown in the console or similarily.

Sopel97 commented:
@Disservin 

no, I mean that stockfish is compiled with `-fno-exceptions`, so any throw is equivalent to `std::terminate`

Disservin commented:
well oh

XInTheDark commented on src/position.cpp:
```diff
@@ -204,87 +202,206 @@ Position& Position::set(const string& fenStr, bool isChess960, StateInfo* si, Th
+      gamePly = 1;
+  }
+
+  // Technically, positions with rule50==100 are correct, just no moves can be made further.
```
This is quite unnecessary IMO, rule50 count doesn't really affect gameplay, and user can easily artifically bypass it anyways.

DemiMarie replied:
While the 50-move rule and 75-move rule do apply in over-the-board play, they do not apply for certain purposes in endgame theory.  In particular, endgame tablebases will allow one to step through a forced with that takes far longer, and this would (presulably) result in a larger count here.  Instead of refusing to import the FEN, I think it would be better for Stockfish to clamp rule50 to 150.

noobpwnftw commented:
Instead of terminating, I suggest you let it whine somewhere **informatively** and refuse to start any search and return `bestmove (null)` as if it was fed with a checkmated position.

vondele commented:
I think that's something to consider.

TheBlackPlague commented:
> Instead of terminating, I suggest you let it whine somewhere **informatively** and refuse to start any search and return `bestmove (null)` as if it was fed with a checkmated position.

Well, the UCI way would be to ignore and not update data internally. So if we have the `startpos` position as the default, UCI would dictate we just don't change it from that. **Completely ignore it.** - Not sure how many people care about that?

Sopel97 commented:
> > Instead of terminating, I suggest you let it whine somewhere **informatively** and refuse to start any search and return `bestmove (null)` as if it was fed with a checkmated position.
> 
> Well, the UCI way would be to ignore and not update data internally. So if we have the `startpos` position as the default, UCI would dictate we just don't change it from that. **Completely ignore it.** - Not sure how many people care about that?

consider

```
position fen rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1 moves e2e4 e7e7 e2e2 e7e5
```

What should be the state right now? Should the whole statment be rolled back after encountering the invalid move? Should it result in `rnbqkbnr/pppp1ppp/8/4p3/4P3/8/PPPP1PPP/RNBQKBNR w KQkq - 0 2`? "ignore" is vague

TheBlackPlague commented:
> > > Instead of terminating, I suggest you let it whine somewhere **informatively** and refuse to start any search and return `bestmove (null)` as if it was fed with a checkmated position.
> > 
> > 
> > Well, the UCI way would be to ignore and not update data internally. So if we have the `startpos` position as the default, UCI would dictate we just don't change it from that. **Completely ignore it.** - Not sure how many people care about that?
> 
> consider
> 
> ```
> position fen rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1 moves e2e4 e7e7 e2e2 e7e5
> ```
> 
> What should be the state right now? Should the whole statment be rolled back after encountering the invalid move? Should it result in `rnbqkbnr/pppp1ppp/8/4p3/4P3/8/PPPP1PPP/RNBQKBNR w KQkq - 0 2`? "ignore" is vague

Per the UCI documentation, the entire statement should be ignored and rolled back.

Sopel97 commented:
> > > > Instead of terminating, I suggest you let it whine somewhere **informatively** and refuse to start any search and return `bestmove (null)` as if it was fed with a checkmated position.
> > > 
> > > 
> > > Well, the UCI way would be to ignore and not update data internally. So if we have the `startpos` position as the default, UCI would dictate we just don't change it from that. **Completely ignore it.** - Not sure how many people care about that?
> > 
> > 
> > consider
> > ```
> > position fen rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1 moves e2e4 e7e7 e2e2 e7e5
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > What should be the state right now? Should the whole statment be rolled back after encountering the invalid move? Should it result in `rnbqkbnr/pppp1ppp/8/4p3/4P3/8/PPPP1PPP/RNBQKBNR w KQkq - 0 2`? "ignore" is vague
> 
> Per the UCI documentation, the entire statement should be ignored and rolled back.

then not even stockfish master is compliant

TheBlackPlague commented:
> > > > > Instead of terminating, I suggest you let it whine somewhere **informatively** and refuse to start any search and return `bestmove (null)` as if it was fed with a checkmated position.
> > > > 
> > > > 
> > > > Well, the UCI way would be to ignore and not update data internally. So if we have the `startpos` position as the default, UCI would dictate we just don't change it from that. **Completely ignore it.** - Not sure how many people care about that?
> > > 
> > > 
> > > consider
> > > ```
> > > position fen rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1 moves e2e4 e7e7 e2e2 e7e5
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > What should be the state right now? Should the whole statment be rolled back after encountering the invalid move? Should it result in `rnbqkbnr/pppp1ppp/8/4p3/4P3/8/PPPP1PPP/RNBQKBNR w KQkq - 0 2`? "ignore" is vague
> > 
> > 
> > Per the UCI documentation, the entire statement should be ignored and rolled back.
> 
> then not even stockfish master is compliant. It gives something completely different still. `rnbqkbnr/pppppppp/8/8/4P3/8/PPPP1PPP/RNBQKBNR b KQkq - 0 1`

Yes, Stockfish `master` isn't compliant per the UCI documentation. I am not sure what else to say. 

I have already called the UCI documentation certain words, and I believe it was a protocol that was thought based on a specific use case and should've never been a widely-used protocol, but here we are. If you wish to be fully compliant, then the UCI protocol clearly states to ignore invalid commands, in which case a complete return to the previous state is the appropriate behavior.

quildtide commented:
> > Instead of terminating, I suggest you let it whine somewhere **informatively** and refuse to start any search and return `bestmove (null)` as if it was fed with a checkmated position.
> 
> Well, the UCI way would be to ignore and not update data internally. So if we have the `startpos` position as the default, UCI would dictate we just don't change it from that. **Completely ignore it.** - Not sure how many people care about that?

A thought experiment: let's say Stockfish decides to use a modified UCI, that let's call "UCI+E")  which is UCI with additional stderr output.

If a client doesn't adopt UCI+E and continues following UCI, UCI+E should function more or less the same (backwards compatibility!), but if a client chooses to adopt it, it could potentially be beneficial.

So if it's useful to be UCI-compliant while also having some kind of sane way of indicating to clients that an input was not okay, then something like this could be good:
- In response to invalid input, Stockfish ignores the whole input (following UCI) and then sends a complaint through stderr.
- Clients that don't adopt this stderr extension to UCI would just experience the proper UCI-compliant response
- Clients that choose to adopt this stderr extension to UCI would be notified that an input was ignored (with possible explanation for why)

TheBlackPlague commented:
> > > Instead of terminating, I suggest you let it whine somewhere **informatively** and refuse to start any search and return `bestmove (null)` as if it was fed with a checkmated position.
> > 
> > 
> > Well, the UCI way would be to ignore and not update data internally. So if we have the `startpos` position as the default, UCI would dictate we just don't change it from that. **Completely ignore it.** - Not sure how many people care about that?
> 
> A thought experiment: let's say Stockfish decides to use a modified UCI, that let's call "UCI+E") which is UCI with additional stderr output.
> 
> If a client doesn't adopt UCI+E and continues following UCI, UCI+E should function more or less the same (backwards compatibility!), but if a client chooses to adopt it, it could potentially be beneficial.
> 
> So if it's useful to be UCI-compliant while also having some kind of sane way of indicating to clients that an input was not okay, then something like this could be good:
> 
> * In response to invalid input, Stockfish ignores the whole input (following UCI) and then sends a complaint through stderr.
> * Clients that don't adopt this stderr extension to UCI would just experience the proper UCI-compliant response
> * Clients that choose to adopt this stderr extension to UCI would be notified that an input was ignored (with possible explanation for why)
> 
> Alternatively, do I understand correctly that a UCI-compliant client is also responsible for ignoring messages from an engine's stdout if it's not something defined in UCI? Complaints could also just be sent through stdout then if that's the case.

Of course. I suggested using an enhanced version of UCI (or one that extends the functionality with numerous other functionality) a long time ago, though, maybe things will be different now.

I believe the right way (& best way), a merge between your and my suggestion, would be a complete rollback (ignorance of the statement in terms of internal data change), but also a good error message detailing why the command was ignored is the right way to go about it. 

What do you think?

ZealanL commented:
I think it is worth noting that restricting the engine to positions with 32 pieces does limit some use cases for it. I've seen many crazy chess puzzles involving more than 32 pieces, and it would be a bit unfortunate to not be able to run Stockfish on those special cases.

TheBlackPlague commented:
> I think it is worth noting that restricting the engine to positions with 16 pieces for each side does limit some use cases for it. I've seen many crazy chess puzzles involving more than 32 pieces, and it would be a bit unfortunate to not be able to run Stockfish on those special cases.

Well, that's the mess you started. 

Regardless, as per @Sopel97's comment:
> inputs that would be experiencing a change would have been already violating the precondition.

This means that the puzzles/challenges you speak of already violate the precondition. So yes, they'll be experiencing a change. This removes undefined behavior, and completely stamps down on the "Stockfish is a **chess engine**, not **imaginary chess engine**."

Consequences of actions.

Sopel97 commented:
> I think it is worth noting that restricting the engine to positions with 16 pieces for each side does limit some use cases for it. I've seen many crazy chess puzzles involving more than 32 pieces, and it would be a bit unfortunate to not be able to run Stockfish on those special cases.

We have no way to guarantee quality analysis for such positions, so it would be dubious even if we supported them. Look at how Stockfish does in matches with odds for example. There are engines better suited for analysing non-chess positions.

FredFergusonPM commented:
> I think it is worth noting that restricting the engine to positions with 16 pieces for each side does limit some use cases for it. I've seen many crazy chess puzzles involving more than 32 pieces, and it would be a bit unfortunate to not be able to run Stockfish on those special cases.

The fallout of all this will be Stockfish will be only for the purists. If anyone wants to do something thats illegal in Standard or FRC, they need to fork and make the changes themselves or use some other alternative.
If anything apart from that is allowed, there will be someone else creating a mess here or reddit or fishcooking or someplace else :)

bftjoe commented:
> I think it is worth noting that restricting the engine to positions with 16 pieces for each side does limit some use cases for it. I've seen many crazy chess puzzles involving more than 32 pieces, and it would be a bit unfortunate to not be able to run Stockfish on those special cases.

Stockfish already can't guarantee quality analysis of such positions due to limitations imposed for performance reasons. Not allowing people to feed such positions to Stockfish and then have them make clickbait YouTube videos about how stockfish is blind is actually a good thing. Use fairy sf or stockfish mv for such positions, anyway.

vdbergh commented:
Probably this comment has been made before. Although UCI is supposed to be stateless, there is still a little bit of state introduced by the `setposition` command. 

So it is seems to be ok that the engine, after receiving a `setposition` command it does not want to handle,  refuses to execute certain commands (e.g. returning `bestmove null` on `go`) until a new valid `setposition` command is received.

People have argued (including me) that UCI mandates that the engine should simply ignore an invalid `setposition` commands. Now I don't support this opinion anymore.

Sopel97 commented:
Moved the actual error handling to the UCI layer. `Position::set` now returns the error. Updated the comment on `Position::set`.

This is close to the final form. Sanity check: https://tests.stockfishchess.org/tests/view/6461f8ae87f6567dd4df2ad8

Sopel97 commented:
> One little issue I have is that the final error message does not seem to say which command is actually the culprit of the error. I would prefer something like Invalid command (setposition): <message> instead of the generic "CRITICAL ERROR".

It does have a somewhat granular error message, though for batched commands it could be even more precise. I'll see if it can be done in a reasonable way.

> Although it seems the idea is regarded as controversial, I would still like to make it possible for the GUI to parse the error message in a simple way if it wants to do this (rather than just relaying the message to the user).

I'd like to see UCI specification first. Then we can implement it in Stockfish.

miguel-l commented:
A little bit nitpicky, but from my limited testing:

-  when compiled using `make build`
```
position fen 7k/7p/8/8/8/8/7P/K6K w - - 0 1 
info string CRITICAL ERROR: Command `position fen 7k/7p/8/8/8/8/7P/K6K w - - 0 1 ` failed. Reason: Invalid FEN.

terminate called without an active exception
Aborted
```

- when compiled using `make build debug=yes`
```
position fen 7k/7p/8/8/8/8/7P/K6K w - - 0 1 
stockfish: position.cpp:1428: bool Stockfish::Position::pos_is_ok() const: Assertion `0 && "pos_is_ok: King count"' failed.
Aborted
```

I think the explicit crash with the `info string` was good but it would be nice to also get a more specific message without recompiling if possible.

Sopel97 commented:
> I think the explicit crash with the info string was good but it would be nice to also get a more specific message without recompiling if possible.

With debug=yes it leaks implementation details and suggests to the user it is a bug in Stockfish.

miguel-l commented:
> > I think the explicit crash with the info string was good but it would be nice to also get a more specific message without recompiling if possible.
> 
> With debug=yes it leaks implementation details and suggests to the user it is a bug in Stockfish.

I agree we shouldn't leak implementation details. I was thinking more of, for example, adding the "king count" part to the info string message rather than "it's in pos.is_ok()". Although I don't immediately see an easy way to change it.

From the user's point of view, it seems inconsistent that some info string errors give details about what's wrong with the position, and some just say it's incorrect.

robertnurnberg commented on src/position.cpp:
```diff
@@ -204,89 +215,208 @@ Position& Position::set(const string& fenStr, bool isChess960, StateInfo* si, Th
+
+      const int wAdditionalKnights = std::max((int)count<KNIGHT>(WHITE) - 2, 0);
+      const int bAdditionalKnights = std::max((int)count<KNIGHT>(BLACK) - 2, 0);
+      const int wAdditionalBishops = std::max((int)count<BISHOP>(WHITE) - 2, 0);
```
Tiny nit: one could rather check difference of number of DSBs and LSBs to 1,  for both white and black. (not sure how easy the number of dark squared bishops can be obtained, for example)

This would help catch a board with two white LSBs, no white DSB and 8 white paws, say.

robertnurnberg commented on src/position.cpp:
```diff
@@ -204,89 +215,208 @@ Position& Position::set(const string& fenStr, bool isChess960, StateInfo* si, Th
+      const int bAdditionalRooks = std::max((int)count<ROOK>(BLACK) - 2, 0);
+      const int wAdditionalQueens = std::max((int)count<QUEEN>(WHITE) - 1, 0);
+      const int bAdditionalQueens = std::max((int)count<QUEEN>(BLACK) - 1, 0);
+      if (wAdditionalKnights + wAdditionalBishops + wAdditionalRooks + wAdditionalQueens > 8 - wPawns)
```
If one defines `8 - wPawns` as `maxPromotions`, the latter could be set to zero if all opposing pawns are still on their original rank. (plus some other cases, where they form an impenetrable wall, that could never have been overcome by an opposing pawn). But these kind of complicated checks may already go beyond the original scope of this PR.

snicolet commented:
So, what's the consensus about this pull request?

vondele commented:
I think we should merge something along these lines, but was holding off until after the SF16 release, and busy afterwards. Probably, this could be revisited now as this gives some time to fix-up eventual fallout.

In the back of my mind, I have been thinking also that there is possibly some overlap with https://github.com/official-stockfish/Stockfish/pull/4626 or said differently, if we move in the direction of having a shared library of SF, we will have to think a bit more about error handling (i.e. a library should not abort).

Disservin commented:
I agree with merging, though we should keep it rather simple in the future otherwise people will constantly try to add new rules to prohibit non legal chess positions. I think all we really care about is avoiding positions where a crash can happen (or which could later cause a crash). 
Btw needs a rebase ; )

DemiMarie commented on src/uci.cpp:
```diff
@@ -395,4 +407,12 @@ Move UCI::to_move(const Position& pos, string& str) {
+  // gives any guarantees as to the program's behaviour for the user to rely on,
+  // so the safest option is to terminate.
+  sync_cout << "info string CRITICAL ERROR: Command `" << fullCommand << "` failed. Reason: " << message << '\n' << sync_endl;
+  std::terminate();
```
Exiting is an appropriate action here, but I don’t think `std::terminate()` is the right way to do it.  `std::terminate()` typically results in a core dump, so it should only be used when there is a bug in Stockfish itself.  It isn’t the appropriate reaction to invalid input.  In this case, I would just call `exit()` if there are no other threads running, or `quick_exit()` or `_Exit` otherwise.

vondele commented:
I'll close this as outdated, even though I think the work on trying to make SF more modular and library-like will require this eventually.
